app/src/main/java/io/github/trojan_gfw/igniter/MainActivity.java

```diff
@@ -44,10 +51,59 @@
     private Switch verifySwitch;
     private Switch clashSwitch;
     private TextView clashLink;
-    private Button startStopButton;
+    private Button startStopButton, saveServerIb;
     private EditText trojanURLText;
+//    private ServerListDialog serverListDialog;
 
     private BroadcastReceiver serviceStateReceiver;
+    private ServerListDataSource serverListDataManager;
+    private TextViewListener remoteAddrTextListener = new TextViewListener() {
+        @Override
+        protected void onTextChanged(String before, String old, String aNew, String after) {
+            // update TextView
+            startUpdates(); // to prevent infinite loop.
+            if (remoteAddrText.hasFocus()) {
+                TrojanConfig ins = Globals.getTrojanConfigInstance();
+                Log.i(TAG, "before: " + before + " old: " + old + " aNew: " + aNew + " after: " + after);
+                Log.i(TAG, "set instance remote addr text: " + remoteAddrText.getText().toString());
+                ins.setRemoteAddr(remoteAddrText.getText().toString());
+            }
+            endUpdates();
+        }
+    };
+    private TextViewListener remotePortTextListener = new TextViewListener() {
+        @Override
+        protected void onTextChanged(String before, String old, String aNew, String after) {
+            // update TextView
+            startUpdates(); // to prevent infinite loop.
+            if (remotePortText.hasFocus()) {
+                TrojanConfig ins = Globals.getTrojanConfigInstance();
+                Log.i(TAG, "set instance remote port text: " + remotePortText.getText().toString());
+                String portStr = remotePortText.getText().toString();
+                try {
+                    int port = Integer.parseInt(portStr);
+                    ins.setRemotePort(port);
+                } catch (NumberFormatException e) {
+                    // Ignore when we get invalid number
+                    e.printStackTrace();
+                }
+            }
+            endUpdates();
+        }
+    };
+    private TextViewListener passwordTextListener = new TextViewListener() {
+        @Override
+        protected void onTextChanged(String before, String old, String aNew, String after) {
+            // update TextView
+            startUpdates(); // to prevent infinite loop.
+            if (passwordText.hasFocus()) {
+                TrojanConfig ins = Globals.getTrojanConfigInstance();
+                Log.i(TAG, "set instance password text: " + passwordText.getText().toString());
```

Comment:
It seems for debug? IMO it is sensitive information.

Reply:
Oops, you're right. I will remove the log at the next commit.